autofs-5.1.8 - dont immediately call function when waiting

From: Ian Kent <raven@themaw.net>

When autofs needs to wait for a sss connection the connection function
is immediately called a second time without first waiting. Adjust the
calling so that there's a wait before the next call.

Signed-off-by: Ian Kent <raven@themaw.net>
---
 CHANGELOG            |    1 +
 modules/lookup_sss.c |   24 ++++++++++++------------
 2 files changed, 13 insertions(+), 12 deletions(-)

--- autofs-5.1.4.orig/CHANGELOG
+++ autofs-5.1.4/CHANGELOG
@@ -107,6 +107,7 @@
 - fix incorrect path for is_mounted() in try_remount().
 - fail on empty replicated host name.
 - improve handling of ENOENT in sss setautomntent().
+- don't immediately call function when waiting.
 
 xx/xx/2018 autofs-5.1.5
 - fix flag file permission.
--- autofs-5.1.4.orig/modules/lookup_sss.c
+++ autofs-5.1.4/modules/lookup_sss.c
@@ -338,10 +338,13 @@ static int setautomntent_wait(unsigned i
 	     "can't connect to sssd, retry for %d seconds",
 	     retries);
 
-	while (++retry <= retries) {
+	while (++retry < retries) {
 		struct timespec t = { SSS_WAIT_INTERVAL, 0 };
 		struct timespec r;
 
+		while (nanosleep(&t, &r) == -1 && errno == EINTR)
+			memcpy(&t, &r, sizeof(struct timespec));
+
 		ret = ctxt->setautomntent(ctxt->mapname, sss_ctxt);
 		if (proto_version(ctxt) == 0) {
 			if (ret != ENOENT)
@@ -355,9 +358,6 @@ static int setautomntent_wait(unsigned i
 			free(*sss_ctxt);
 			*sss_ctxt = NULL;
 		}
-
-		while (nanosleep(&t, &r) == -1 && errno == EINTR)
-			memcpy(&t, &r, sizeof(struct timespec));
 	}
 
 	if (!ret)
@@ -475,10 +475,13 @@ static int getautomntent_wait(unsigned i
 	 "can't contact sssd to to get map entry, retry for %d seconds",
 	 retries);
 
-	while (++retry <= retries) {
+	while (++retry < retries) {
 		struct timespec t = { SSS_WAIT_INTERVAL, 0 };
 		struct timespec r;
 
+		while (nanosleep(&t, &r) == -1 && errno == EINTR)
+			memcpy(&t, &r, sizeof(struct timespec));
+
 		ret = ctxt->getautomntent_r(key, value, sss_ctxt);
 		if (proto_version(ctxt) == 0) {
 			if (ret != ENOENT)
@@ -487,9 +490,6 @@ static int getautomntent_wait(unsigned i
 			if (ret != EHOSTDOWN)
 				break;
 		}
-
-		while (nanosleep(&t, &r) == -1 && errno == EINTR)
-			memcpy(&t, &r, sizeof(struct timespec));
 	}
 
 	if (!ret)
@@ -600,10 +600,13 @@ static int getautomntbyname_wait(unsigne
 	"can't contact sssd to to lookup key value, retry for %d seconds",
 	retries);
 
-	while (++retry <= retries) {
+	while (++retry < retries) {
 		struct timespec t = { SSS_WAIT_INTERVAL, 0 };
 		struct timespec r;
 
+		while (nanosleep(&t, &r) == -1 && errno == EINTR)
+			memcpy(&t, &r, sizeof(struct timespec));
+
 		ret = ctxt->getautomntbyname_r(key, value, sss_ctxt);
 		if (proto_version(ctxt) == 0) {
 			if (ret != ENOENT)
@@ -612,9 +615,6 @@ static int getautomntbyname_wait(unsigne
 			if (ret != EHOSTDOWN)
 				break;
 		}
-
-		while (nanosleep(&t, &r) == -1 && errno == EINTR)
-			memcpy(&t, &r, sizeof(struct timespec));
 	}
 
 	if (!ret)
